File: Recursion_and_Backtracking_pepcoding_level_1/1.Recursion_Intro/5.power_v1.js
const power_linear = (x, n) => {
    if (n === 0) return 1;
    let curr = x;
    let child = power_linear(x, n - 1);
    return curr * child;
};

console.log(power_linear(2, 6));
/////////////////////////////////////////////////////////

const power_logarithmic = (x, n) => {
    if (n === 0) return 1;

    let x2 = power_logarithmic(x, n / 2);
    let xn = x2 * x2;
    if (n & 1) xn = xn * x;

    return xn;
};

console.log(power_logarithmic(2, 6));

File: Recursion_and_Backtracking_pepcoding_level_1/1.Recursion_Intro/5.power_v1.rb
def power_linear x, n
  return 1 if n == 0

  return x * power_linear(x, n-1)
end

puts power_linear 3, 5



def power_logarithmic x, n
    return 1 if n == 0

    xn2 = power_logarithmic x, n/2

    return xn2 * xn2 if n.even?
    return x * xn2 * xn2 if n.odd?
end

puts power_logarithmic 3, 5
